File: agent/tsdb-cloudflare/src/storage/column-encoder.ts
/**
 * Encodes and decodes individual column data blocks.
 *
 * Each encode* returns a deflate-compressed Uint8Array; decode* is the inverse.
 *
 * Encoding strategies:
 *   Timestamp      — delta encode (first absolute, rest forward-deltas), Int64 LE, deflate.
 *   String         — dictionary encoded (dict + int32 indices), deflate.
 *   NullableString — null bitmap + dictionary (nulls → sentinel index), deflate.
 *   NullableDouble — null bitmap + non-null doubles, deflate.
 *   Byte           — raw bytes, deflate.
 */

import { compress, decompress } from "../lib/compression";
import { BloomFilter, buildBloom } from "./bloom-filter";

const ENCODER = new TextEncoder();
const DECODER = new TextDecoder();

// Sentinel for null in nullable-string indices (matches .NET int.MinValue)
const NULL_SENTINEL = -2147483648;

// ── Timestamps ────────────────────────────────────────────────────────────────

/**
 * Delta-encode timestamps as Int64 LE and compress.
 * First value is absolute; remaining are forward deltas.
 */
export async function encodeTimestamps(values: number[]): Promise<Uint8Array> {
  if (values.length === 0) return new Uint8Array(0);

  const buf = new ArrayBuffer(values.length * 8);
  const view = new DataView(buf);

  view.setBigInt64(0, BigInt(values[0]), true);
  for (let i = 1; i < values.length; i++) {
    view.setBigInt64(i * 8, BigInt(values[i] - values[i - 1]), true);
  }

  return compress(new Uint8Array(buf));
}

export async function decodeTimestamps(
  compressed: Uint8Array,
  count: number,
): Promise<number[]> {
  const raw = await decompress(compressed);
  const view = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  const result = new Array<number>(count);
  let prev = 0;

  for (let i = 0; i < count; i++) {
    prev += Number(view.getBigInt64(i * 8, true));
    result[i] = prev;
  }

  return result;
}

// ── Strings (non-nullable) ────────────────────────────────────────────────────

export async function encodeStrings(
  values: string[],
  doBuildBloom = false,
): Promise<{ encoded: Uint8Array; bloom: BloomFilter | null }> {
  const { dictList, indices } = buildDictionary(values);
  const bloom = doBuildBloom ? buildBloom(values) : null;
  const serialized = serializeDictEncoded(dictList, indices, null);
  return { encoded: await compress(serialized), bloom };
}

export async function decodeStrings(
  compressed: Uint8Array,
  count: number,
): Promise<string[]> {
  const raw = await decompress(compressed);
  const reader = new BinaryReader(raw);

  const dict = readDict(reader);
  const result = new Array<string>(count);
  for (let i = 0; i < count; i++) {
    result[i] = dict[reader.readInt32()];
  }
  return result;
}

// ── Nullable strings ──────────────────────────────────────────────────────────

export async function encodeNullableStrings(
  values: (string | null)[],
  doBuildBloom = false,
): Promise<{ encoded: Uint8Array; bloom: BloomFilter | null }> {
  const { bitmap, nonNulls } = buildNullBitmap(values);
  const { dictList, indices } = buildNullableDictionary(values);

  const bloom =
    doBuildBloom && nonNulls.length > 0 ? buildBloom(nonNulls) : null;

  const serialized = serializeDictEncoded(dictList, indices, bitmap);
  return { encoded: await compress(serialized), bloom };
}

export async function decodeNullableStrings(
  compressed: Uint8Array,
  count: number,
): Promise<(string | null)[]> {
  const raw = await decompress(compressed);
  const reader = new BinaryReader(raw);

  const bitmap = readNullBitmap(reader);
  const dict = readDict(reader);
  const result = new Array<string | null>(count);

  for (let i = 0; i < count; i++) {
    const idx = reader.readInt32();
    result[i] = idx === NULL_SENTINEL ? null : dict[idx];
  }
  return result;
}

// ── Nullable doubles ──────────────────────────────────────────────────────────

export async function encodeNullableDoubles(
  values: (number | null)[],
): Promise<Uint8Array> {
  const bitmapLen = (values.length + 7) >> 3;
  const bitmap = new Uint8Array(bitmapLen);
  const nonNulls: number[] = [];

  for (let i = 0; i < values.length; i++) {
    if (values[i] === null) {
      bitmap[i >> 3] |= 1 << (i & 7);
    } else {
      nonNulls.push(values[i]!);
    }
  }

  // bitmapLen(i32) + bitmap + nonNullCount(i32) + doubles
  const totalBytes = 4 + bitmapLen + 4 + nonNulls.length * 8;
  const buf = new ArrayBuffer(totalBytes);
  const view = new DataView(buf);
  const bytes = new Uint8Array(buf);
  let offset = 0;

  view.setInt32(offset, bitmapLen, true);
  offset += 4;
  bytes.set(bitmap, offset);
  offset += bitmapLen;
  view.setInt32(offset, nonNulls.length, true);
  offset += 4;
  for (const d of nonNulls) {
    view.setFloat64(offset, d, true);
    offset += 8;
  }

  return compress(new Uint8Array(buf));
}

export async function decodeNullableDoubles(
  compressed: Uint8Array,
  count: number,
): Promise<(number | null)[]> {
  const raw = await decompress(compressed);
  const view = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  let offset = 0;

  const bitmapLen = view.getInt32(offset, true);
  offset += 4;
  const bitmap = raw.slice(offset, offset + bitmapLen);
  offset += bitmapLen;

  const nonNullCount = view.getInt32(offset, true);
  offset += 4;
  const nonNulls = new Array<number>(nonNullCount);
  for (let i = 0; i < nonNullCount; i++) {
    nonNulls[i] = view.getFloat64(offset, true);
    offset += 8;
  }

  const result = new Array<number | null>(count);
  let nni = 0;
  for (let i = 0; i < count; i++) {
    result[i] =
      (bitmap[i >> 3] & (1 << (i & 7))) !== 0 ? null : nonNulls[nni++];
  }
  return result;
}

// ── Raw bytes ─────────────────────────────────────────────────────────────────

export async function encodeBytes(values: Uint8Array): Promise<Uint8Array> {
  return compress(values);
}

export async function decodeBytes(
  compressed: Uint8Array,
  _count: number,
): Promise<Uint8Array> {
  return decompress(compressed);
}

// ── Private helpers ───────────────────────────────────────────────────────────

function buildDictionary(values: string[]): {
  dictList: string[];
  indices: Int32Array;
} {
  const dict = new Map<string, number>();
  const dictList: string[] = [];
  const indices = new Int32Array(values.length);

  for (let i = 0; i < values.length; i++) {
    const s = values[i];
    let idx = dict.get(s);
    if (idx === undefined) {
      idx = dictList.length;
      dict.set(s, idx);
      dictList.push(s);
    }
    indices[i] = idx;
  }
  return { dictList, indices };
}

function buildNullableDictionary(values: (string | null)[]): {
  dictList: string[];
  indices: Int32Array;
} {
  const dict = new Map<string, number>();
  const dictList: string[] = [];
  const indices = new Int32Array(values.length);

  for (let i = 0; i < values.length; i++) {
    const s = values[i];
    if (s === null) {
      indices[i] = NULL_SENTINEL;
      continue;
    }
    let idx = dict.get(s);
    if (idx === undefined) {
      idx = dictList.length;
      dict.set(s, idx);
      dictList.push(s);
    }
    indices[i] = idx;
  }
  return { dictList, indices };
}

function buildNullBitmap(values: (string | null)[]): {
  bitmap: Uint8Array;
  nonNulls: string[];
} {
  const bitmap = new Uint8Array((values.length + 7) >> 3);
  const nonNulls: string[] = [];

  for (let i = 0; i < values.length; i++) {
    if (values[i] === null) {
      bitmap[i >> 3] |= 1 << (i & 7);
    } else {
      nonNulls.push(values[i]!);
    }
  }
  return { bitmap, nonNulls };
}

/**
 * Serialize dict-encoded column:
 *   [nullBitmap?] → bitmapLen(i32) + bitmap(bytes)
 *   [dict]        → dictCount(i32) + for each: byteLen(i32) + utf8
 *   [indices]     → int32 per row
 */
function serializeDictEncoded(
  dictList: string[],
  indices: Int32Array,
  nullBitmap: Uint8Array | null,
): Uint8Array {
  // Pre-encode all dict strings to compute total size
  const encodedStrings = dictList.map((s) => ENCODER.encode(s));
  let totalSize = 0;

  // Null bitmap section
  if (nullBitmap !== null) {
    totalSize += 4 + nullBitmap.length; // bitmapLen + bitmap bytes
  }

  // Dict section: count + (len + bytes) per entry
  totalSize += 4; // dictCount
  for (const enc of encodedStrings) {
    totalSize += 4 + enc.length;
  }

  // Indices section
  totalSize += indices.length * 4;

  const buf = new ArrayBuffer(totalSize);
  const view = new DataView(buf);
  const bytes = new Uint8Array(buf);
  let offset = 0;

  // Null bitmap
  if (nullBitmap !== null) {
    view.setInt32(offset, nullBitmap.length, true);
    offset += 4;
    bytes.set(nullBitmap, offset);
    offset += nullBitmap.length;
  }

  // Dictionary
  view.setInt32(offset, dictList.length, true);
  offset += 4;
  for (const enc of encodedStrings) {
    view.setInt32(offset, enc.length, true);
    offset += 4;
    bytes.set(enc, offset);
    offset += enc.length;
  }

  // Indices
  for (let i = 0; i < indices.length; i++) {
    view.setInt32(offset, indices[i], true);
    offset += 4;
  }

  return new Uint8Array(buf);
}

// ── BinaryReader utility ──────────────────────────────────────────────────────

/** Minimal forward-only reader over a Uint8Array, matching .NET BinaryReader semantics. */
class BinaryReader {
  private readonly view: DataView;
  private readonly bytes: Uint8Array;
  private pos = 0;

  constructor(data: Uint8Array) {
    this.bytes = data;
    this.view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  }

  readInt32(): number {
    const v = this.view.getInt32(this.pos, true);
    this.pos += 4;
    return v;
  }

  readBytes(count: number): Uint8Array {
    const slice = this.bytes.slice(this.pos, this.pos + count);
    this.pos += count;
    return slice;
  }
}

function readDict(reader: BinaryReader): string[] {
  const count = reader.readInt32();
  const dict = new Array<string>(count);
  for (let i = 0; i < count; i++) {
    const len = reader.readInt32();
    const bytes = reader.readBytes(len);
    dict[i] = DECODER.decode(bytes);
  }
  return dict;
}

function readNullBitmap(reader: BinaryReader): Uint8Array {
  const len = reader.readInt32();
  return reader.readBytes(len);
}

File: agent/tsdb/src/FeatBit.DataWarehouse/Storage/ColumnEncoder.cs
using System.Buffers;
using System.Buffers.Binary;
using System.IO.Compression;
using System.Text;

namespace FeatBit.DataWarehouse.Storage;

/// <summary>
/// Encodes and decodes individual column data blocks.
///
/// Each Encode* method returns a Brotli-compressed byte array ready to be
/// written directly into the segment file.  Decode* is the symmetric inverse.
///
/// Encoding strategies by type:
///   Timestamp      — delta encode (first value absolute, rest forward-deltas),
///                    pack as little-endian int64[], Brotli compress.
///   String         — build a string dictionary; store (dict + int32[] indices), Brotli.
///   NullableString — null bitmap + same as String (nulls get index = int.MinValue).
///   NullableDouble — null bitmap + raw double[] for non-null values, Brotli.
///   Byte           — raw byte[], Brotli.
/// </summary>
internal static class ColumnEncoder
{
    // ── Timestamps ────────────────────────────────────────────────────────────

    public static byte[] EncodeTimestamps(long[] values)
    {
        if (values.Length == 0) return [];

        var raw = new byte[values.Length * sizeof(long)];
        var span = raw.AsSpan();

        // First value is absolute; rest are forward deltas.
        BinaryPrimitives.WriteInt64LittleEndian(span, values[0]);
        for (int i = 1; i < values.Length; i++)
            BinaryPrimitives.WriteInt64LittleEndian(span[(i * 8)..], values[i] - values[i - 1]);

        return BrotliCompress(raw);
    }

    public static long[] DecodeTimestamps(byte[] compressed, int count)
    {
        var raw    = BrotliDecompress(compressed, count * sizeof(long));
        var result = new long[count];
        long prev  = 0;

        for (int i = 0; i < count; i++)
        {
            prev      += BinaryPrimitives.ReadInt64LittleEndian(raw.AsSpan(i * 8));
            result[i]  = prev;
        }

        return result;
    }

    // ── Strings (non-nullable) ────────────────────────────────────────────────

    /// <returns>Compressed column bytes and an optional bloom filter (pass buildBloom=true for indexed columns).</returns>
    public static (byte[] Encoded, BloomFilter? Bloom) EncodeStrings(
        string[] values, bool buildBloom = false)
    {
        var (dict, dictList, indices) = BuildDictionary(values);

        var bloom = buildBloom ? BuildBloom(values) : null;

        return (BrotliCompress(SerializeDictEncoded(dictList, indices, nullBitmap: null)), bloom);
    }

    public static string[] DecodeStrings(byte[] compressed, int count)
    {
        var raw  = BrotliDecompress(compressed, 0);
        using var br = new BinaryReader(new MemoryStream(raw), Encoding.UTF8, leaveOpen: false);

        var dict = ReadDict(br);
        var result = new string[count];
        for (int i = 0; i < count; i++)
            result[i] = dict[br.ReadInt32()];

        return result;
    }

    // ── Nullable strings ──────────────────────────────────────────────────────

    public static (byte[] Encoded, BloomFilter? Bloom) EncodeNullableStrings(
        string?[] values, bool buildBloom = false)
    {
        var (nullBitmap, nonNulls) = BuildNullBitmap(values);
        var (_, dictList, indices) = BuildDictionary(nonNullValues: values, nullable: true);

        var bloom = buildBloom && nonNulls.Count > 0
            ? BuildBloom(nonNulls)
            : null;

        return (BrotliCompress(SerializeDictEncoded(dictList, indices, nullBitmap)), bloom);
    }

    public static string?[] DecodeNullableStrings(byte[] compressed, int count)
    {
        var raw = BrotliDecompress(compressed, 0);
        using var br = new BinaryReader(new MemoryStream(raw), Encoding.UTF8, leaveOpen: false);

        var bitmap    = ReadNullBitmap(br);
        var dict      = ReadDict(br);
        var result    = new string?[count];

        for (int i = 0; i < count; i++)
        {
            var idx = br.ReadInt32();
            result[i] = idx == int.MinValue ? null : dict[idx];
        }

        return result;
    }

    // ── Nullable doubles ──────────────────────────────────────────────────────

    public static byte[] EncodeNullableDoubles(double?[] values)
    {
        var bitmapLen = (values.Length + 7) / 8;
        var bitmap    = new byte[bitmapLen];
        var nonNulls  = new List<double>(values.Length);

        for (int i = 0; i < values.Length; i++)
        {
            if (!values[i].HasValue)
                bitmap[i >> 3] |= (byte)(1 << (i & 7));
            else
                nonNulls.Add(values[i]!.Value);
        }

        using var ms = new MemoryStream();
        using var bw = new BinaryWriter(ms, Encoding.UTF8, leaveOpen: true);

        bw.Write(bitmapLen);
        bw.Write(bitmap);
        bw.Write(nonNulls.Count);
        foreach (var d in nonNulls) bw.Write(d);
        bw.Flush();

        return BrotliCompress(ms.ToArray());
    }

    public static double?[] DecodeNullableDoubles(byte[] compressed, int count)
    {
        var raw = BrotliDecompress(compressed, 0);
        using var br = new BinaryReader(new MemoryStream(raw), Encoding.UTF8, leaveOpen: false);

        var bitmapLen   = br.ReadInt32();
        var bitmap      = br.ReadBytes(bitmapLen);
        var nonNullCount = br.ReadInt32();
        var nonNulls    = new double[nonNullCount];
        for (int i = 0; i < nonNullCount; i++) nonNulls[i] = br.ReadDouble();

        var result      = new double?[count];
        int nonNullIdx  = 0;
        for (int i = 0; i < count; i++)
            result[i] = (bitmap[i >> 3] & (1 << (i & 7))) != 0
                ? null
                : nonNulls[nonNullIdx++];

        return result;
    }

    // ── Raw bytes ─────────────────────────────────────────────────────────────

    public static byte[]  EncodeBytes(byte[] values)                  => BrotliCompress(values);
    public static byte[]  DecodeBytes(byte[] compressed, int count)   => BrotliDecompress(compressed, count);

    // ── Brotli helpers ────────────────────────────────────────────────────────

    internal static byte[] BrotliCompress(byte[] data)
    {
        if (data.Length == 0) return [];

        var maxLen = BrotliEncoder.GetMaxCompressedLength(data.Length);
        var buf    = new byte[maxLen];

        // Quality 4 = fast compression, good ratio for time-series columns.
        if (!BrotliEncoder.TryCompress(data, buf, out var written, quality: 4, window: 22))
            throw new InvalidOperationException("BrotliEncoder.TryCompress returned false.");

        return buf[..written];
    }

    internal static byte[] BrotliDecompress(byte[] compressed, int expectedSize)
    {
        if (compressed.Length == 0) return [];

        // Start with expectedSize (if known) or 4× compressed size; double on DestinationTooSmall.
        var bufSize = expectedSize > 0 ? expectedSize : compressed.Length * 4;

        while (true)
        {
            var buf     = new byte[bufSize];
            bool ok     = BrotliDecoder.TryDecompress(compressed, buf, out var written);

            if (ok) return buf[..written];

            // Buffer too small — double and retry.
            bufSize = checked(bufSize * 2);
            if (bufSize > 512 * 1024 * 1024)
                throw new InvalidDataException("Decompressed column exceeds 512 MB or data is invalid.");
        }
    }

    // ── Private helpers ───────────────────────────────────────────────────────

    /// <summary>
    /// Build a string dictionary.  For nullable callers, pass nullable=true and use the
    /// overload that accepts string?[]; nulls receive index = int.MinValue.
    /// </summary>
    private static (Dictionary<string, int> Dict, List<string> DictList, int[] Indices)
        BuildDictionary(string[] values)
    {
        var dict     = new Dictionary<string, int>(values.Length);
        var dictList = new List<string>(values.Length);
        var indices  = new int[values.Length];

        for (int i = 0; i < values.Length; i++)
        {
            var s = values[i];
            if (!dict.TryGetValue(s, out var idx))
            {
                idx      = dictList.Count;
                dict[s]  = idx;
                dictList.Add(s);
            }
            indices[i] = idx;
        }

        return (dict, dictList, indices);
    }

    private static (Dictionary<string, int> Dict, List<string> DictList, int[] Indices)
        BuildDictionary(string?[] nonNullValues, bool nullable)
    {
        var dict     = new Dictionary<string, int>();
        var dictList = new List<string>();
        var indices  = new int[nonNullValues.Length];

        for (int i = 0; i < nonNullValues.Length; i++)
        {
            var s = nonNullValues[i];
            if (s is null)
            {
                indices[i] = int.MinValue; // null sentinel
                continue;
            }
            if (!dict.TryGetValue(s, out var idx))
            {
                idx     = dictList.Count;
                dict[s] = idx;
                dictList.Add(s);
            }
            indices[i] = idx;
        }

        return (dict, dictList, indices);
    }

    private static (byte[] Bitmap, List<string> NonNulls) BuildNullBitmap(string?[] values)
    {
        var bitmap   = new byte[(values.Length + 7) / 8];
        var nonNulls = new List<string>(values.Length);

        for (int i = 0; i < values.Length; i++)
        {
            if (values[i] is null)
                bitmap[i >> 3] |= (byte)(1 << (i & 7));
            else
                nonNulls.Add(values[i]!);
        }

        return (bitmap, nonNulls);
    }

    private static byte[] SerializeDictEncoded(List<string> dictList, int[] indices, byte[]? nullBitmap)
    {
        using var ms = new MemoryStream();
        using var bw = new BinaryWriter(ms, Encoding.UTF8, leaveOpen: true);

        // Null bitmap (only for nullable columns)
        if (nullBitmap is not null)
        {
            bw.Write(nullBitmap.Length);
            bw.Write(nullBitmap);
        }

        // Dictionary
        bw.Write(dictList.Count);
        foreach (var s in dictList)
        {
            var bytes = Encoding.UTF8.GetBytes(s);
            bw.Write(bytes.Length);
            bw.Write(bytes);
        }

        // Indices (int.MinValue = null sentinel for nullable columns)
        foreach (var idx in indices)
            bw.Write(idx);

        bw.Flush();
        return ms.ToArray();
    }

    private static string[] ReadDict(BinaryReader br)
    {
        var count = br.ReadInt32();
        var dict  = new string[count];
        for (int i = 0; i < count; i++)
        {
            var len   = br.ReadInt32();
            var bytes = br.ReadBytes(len);
            dict[i]   = Encoding.UTF8.GetString(bytes);
        }
        return dict;
    }

    private static byte[] ReadNullBitmap(BinaryReader br)
    {
        var len = br.ReadInt32();
        return br.ReadBytes(len);
    }

    private static BloomFilter BuildBloom(IEnumerable<string> values)
    {
        var list  = values as ICollection<string> ?? values.ToList();
        var bloom = new BloomFilter(list.Count);
        foreach (var s in list) bloom.Add(s);
        return bloom;
    }

    private static BloomFilter BuildBloom(List<string> values)
    {
        var bloom = new BloomFilter(values.Count);
        foreach (var s in values) bloom.Add(s);
        return bloom;
    }
}
